src/tiled/tilesetdock.cpp

```diff
@@ -939,3 +942,29 @@ void TilesetDock::swapTiles(Tile *tileA, Tile *tileB)
     QUndoStack *undoStack = mMapDocument->undoStack();
     undoStack->push(new SwapTiles(mMapDocument, tileA, tileB));
 }
+
+static QList<MapObject*> chooseTileObjects(QList<MapObject*> objects)
+{
+    QList<MapObject*> tileObjects;
+
+    for (auto object : objects)
+        if (!object->cell().isEmpty() && object->cell().tile() != nullptr)
```

Comment:
You could simplify this check to just `if (object->cell().tile())`, however I think it's not the behavior we want. I think this code should check only `if (!object->cell().isEmpty()`, since when the tile happens to be set but is invalid, we still want to allow the user to replace it with another tile.

One problem with that though is undo, which should then restore the Tileset* and (invalid) tile ID, rather than setting the tile explicitly to null. So I think in `ChangeMapObjectsTile` you'll need to replace `mOldTiles` with `QVector<Cell> mOldCells`. And of course that code can't use the Tile* unconditionally.